delta/plugins/project-deletion/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/projectdeletion/ProjectDeletionPlugin.scala

```diff
@@ -0,0 +1,164 @@
+import fs2.Stream
+import monix.bio.{Fiber, IO, Task, UIO}
+import cats.syntax.all._
+import ch.epfl.bluebrain.nexus.delta.sourcing.model.Identity.Subject
+
+import java.time.{Duration, Instant}
+import scala.concurrent.duration.DurationInt
+
+/**
+  * ProjectDeletion process reference, also the plugin instance that allows for graceful stop. Once the instance is
+  * constructed, the process would be already running.
+  *
+  * @param fiber
+  *   the underlying process fiber which controls the process
+  */
+class ProjectDeletionPlugin private (fiber: Fiber[Throwable, Unit]) extends Plugin {
+
+  private val logger: Logger = Logger[ProjectDeletionPlugin]
+  logger.info("Project Deletion periodic check has started.")
+
+  override def stop(): Task[Unit] = fiber.cancel.timeout(10.seconds).flatMap {
+    case Some(()) => UIO.delay(logger.info("Project Deletion plugin stopped."))
+    case None     => UIO.delay(logger.error("Project Deletion plugin could not be stopped in 10 seconds."))
+  }
+}
+
+object ProjectDeletionPlugin {
+  implicit private class BooleanTaskOps(val left: Boolean) extends AnyVal {
+    def `<||>`(right: Task[Boolean]): Task[Boolean] = {
+      if (left) {
+        Task.pure(true)
+      } else {
+        right
+      }
+    }
+
+    def `<&&>`(right: Task[Boolean]): Task[Boolean] = {
+      if (left) {
+        right
+      } else {
+        Task.pure(false)
+      }
+    }
+  }
+
+  def projectDeletionPass(
+      allProjects: UIO[Seq[ProjectResource]],
+      deleteProject: ProjectResource => IO[ProjectRejection, Unit],
+      config: ProjectDeletionConfig,
+      lastEventTime: ProjectResource => Task[Instant]
+  ): Task[Unit] = {
+
+    def isIncluded(pr: ProjectResource): Boolean = {
+      config.includedProjects.exists(regex => regex.matches(pr.value.ref.toString))
+    }
+
+    def notExcluded(pr: ProjectResource): Boolean = {
+      !config.excludedProjects.exists(regex => regex.matches(pr.value.ref.toString))
+    }
+
+    def deletableDueToDeprecation(pr: ProjectResource): Boolean = {
+      config.deleteDeprecatedProjects && pr.deprecated
+    }
+
+    def deletableDueToBeingIdle(pr: ProjectResource, now: Instant): Task[Boolean] = {
+      projectIsIdle(pr, now) <&&> resourcesAreIdle(pr, now)
+    }
+
+    def projectIsIdle(pr: ProjectResource, now: Instant) = {
+      (now diff pr.updatedAt).toSeconds > config.idleInterval.toSeconds
+    }
+
+    def resourcesAreIdle(pr: ProjectResource, now: Instant) = {
+      lastEventTime(pr).map(_.isBefore(now.minus(Duration.ofMillis(config.idleInterval.toMillis))))
+    }
+
+    def allProjectsNotAlreadyDeleted: IO[Nothing, Seq[ProjectResource]] = {
+      allProjects
+        .map(_.filter(!_.value.markedForDeletion))
+    }
+
+    def shouldBeDeleted(now: Instant): ProjectResource => Task[Boolean] = { pr =>
+      {
+        (isIncluded(pr) && notExcluded(pr)) <&&> (deletableDueToDeprecation(pr) <||> deletableDueToBeingIdle(pr, now))
```

Comment:
If all the conditions returns a `UIO[Boolean]`,  cats can help there with something like (the or condition can be handled before):
```
implicit val semigroup: Semigroup[Boolean] = (x: Boolean, y: Boolean) => x && y
val isCandidate = (isIncluded(pr), notExcluded(pr), ...).reduce
```

Reply:
The problem here is that we have && and || behaving differently, however both would be viable semigroup implementations for boolean. I think the solution here is clear, although I do agree it should be `UIO[Boolean]`

Reply:
I know but the idea is to combine the or before with:
```
val idleOrDeprecated = ???
val isCandidate = (isIncluded(pr), notExcluded(pr), ..., idleOrDeprecated).reduce
```

Reply:
The options are:

`val intermediate = (deletableDueToDeprecation(pr), deletableDueToBeingIdle(pr, now)).reduce(orSemigroup)
(isIncluded(pr), notExcluded(pr), !alreadyDeleted(pr), intermediate).reduce(andSemigroup)`

or

`(isIncluded(pr), notExcluded(pr), !alreadyDeleted(pr), (deletableDueToDeprecation(pr), deletableDueToBeingIdle(pr, now)).reduce(orSemigroup)).reduce(andSemigroup)`

Both of these seem much less readable than the original. Any suggestions for improvement?

Reply:
I would go with the first one without introducing a semigroup for the `or` part as it is a bit overkill for only two conditions

Reply:
There needs to be a semigroup for the or because one of them is a UIO

If I didn't have an or semigroup then the semigroup logic would just be inline, which would be even worse I think

IMO the custom operators much more readable than these options, I would rather not change to this

Imagine if we need to change the logic to be more complicated. It's going to be much more difficult if we are are using reduce because the reduces will start getting nested

Reply:
The semigroup is just a way to combine things, there are other alternatives for only two conditions (keep the or semigroup if you want) and we really need a good reason to introduce custom operators.
Semigroups is something we all need to understand.
The logic is not going to get more complicated (we want the conditions for deleting a project easy enough to understand to avoid accidents)

Reply:
I understand perfectly how semigroups work, but that doesn't change that when presented in terms of semigroups, this simple boolean expression becomes much more difficult to read. I don't see why making a function that encapsulates a really simple concept and enhances readability / changeability is a bad thing worthy of not merging this PR.

I will make the change in order to get this PR merged